src/runtime/recipe/handle.ts

```diff
@@ -24,6 +24,26 @@ import {ClaimIsTag, Claim} from '../particle-claim.js';
 import {StorageKey} from '../storageNG/storage-key.js';
 import {Capabilities} from '../capabilities.js';
 
+// Represents a relationship between handles.
+// The only example right now is a join handle associating a number
+// of regular handles to expose their join.
+// E.g.
+//   foo: join (bar, baz)
+// Would result in 2 handle associations:
+//   foo (associator) ~ bar (associate)
+//   foo (associator) ~ baz (associate)
+export class HandleAssociation {
+  constructor(
+    // The handle being associated, e.g. a 'map' handle that is joined.
+    public readonly associate: Handle,
+    // The handle associating other handle, e.g. a 'join' handle exposing a join of data.
+    public readonly associator: Handle,
+    // The type of the association.
+    // 'associate' is being "read" as [T]
+    // 'associator' is being "written to" as [(..., &T, ...)]
+    public readonly type: Type) {}
```

Comment:
is Type ever divergent from associate's type? If not, then could we avoid complexity by having a list of associated handles directly on the associator?

Reply:
If associate doesn't have a link to the associator, then when time comes to resolve its type we don't have the constraint that comes from a join.

Have you seen `it('can resolve a recipe with a synthetic join handle and multiple readers'` ? It all works out nicely, multiple particles can read different things from the join as well as mapped handles, and the types get worked out properly by the type system.

Given I have this working properly (I think!) I would prefer not to dumb it down with restrictions that some types have to match if it's not the case in the longer run.

Reply:
the link can be bidirectional; I'm querying whether we need an additional object.

Reply:
Ok, you were correct. I did manage to get this to work without the additional type variable (which was the only thing that was hanging off this relationship object) and having the same type being the type of a handle and the type exposed in the join.

Good job pushing me to simplify this.

Reply:
Thanks for bearing with me! This looks super neat now.